src/hamcrest/core/core/future.py

```diff
@@ -0,0 +1,138 @@
+import sys
+import re
+import asyncio
+from typing import (
+    Any,
+    Optional,
+    Type,
+    TypeVar,
+    Union,
+    Generator,
+    Awaitable,
+)
+
+from hamcrest.core.base_matcher import BaseMatcher
+from hamcrest.core.description import Description
+from hamcrest.core.matcher import Matcher
+
+__author__ = "David Keijser"
+__copyright__ = "Copyright 2021 hamcrest.org"
+__license__ = "BSD, see License.txt"
+
+T = TypeVar("T")
+
+if sys.version_info > (3, 9):
+    # Same as used for asyncio.ensure_future
+    FutureT = Union[asyncio.Future[T], Generator[Any, None, T], Awaitable[T]]
+else:
+    # Future is not a parametrised type in earlier version of python
+    FutureT = Union[asyncio.Future, Generator, Awaitable]
+
+
+class FutureException(BaseMatcher[asyncio.Future]):
+    def __init__(
+        self,
+        expected: Type[Exception],
+        pattern: Optional[str] = None,
+        matching: Optional[Matcher] = None,
+    ) -> None:
+        self.pattern = pattern
+        self.matcher = matching
+        self.expected = expected
+
+    def _matches(self, future: asyncio.Future) -> bool:
+        if not asyncio.isfuture(future):
+            return False
+
+        if not future.done():
+            return False
+
+        if future.cancelled():
+            return False
+
+        exc = future.exception()
+        if exc is None:
+            return False
+
+        if isinstance(exc, self.expected):
+            if self.pattern is not None:
+                if re.search(self.pattern, str(exc)) is None:
+                    return False
+            if self.matcher is not None:
+                if not self.matcher.matches(exc):
+                    return False
+            return True
+
+        return False
+
+    def describe_to(self, description: Description) -> None:
+        description.append_text("Expected a future with exception %s" % self.expected)
```

Comment:
If I'm reading the above right, it'd be more true to say "Expected a _completed_ future ..." etc, right?